broker/supervisor.go

```diff
@@ -0,0 +1,98 @@
+package broker
+
+import (
+	"fmt"
+	"sync"
+)
+
+type Supervisor struct {
+	IsRunning              bool
+	clientCount            uint
+	clientsMutex           *sync.Mutex
+	clientsMapping         map[string]*Client
+	clientOperationChannel chan ClientOperation
+	stoppedSignal          chan struct{}
+	startedSignal          chan struct{}
+	mainWg                 *sync.WaitGroup
+}
+
+func NewSupervisor() *Supervisor {
+	return &Supervisor{
+		clientCount:            0,
+		IsRunning:              false,
+		clientsMutex:           &sync.Mutex{},
+		clientsMapping:         make(map[string]*Client),
+		clientOperationChannel: make(chan ClientOperation),
+		stoppedSignal:          make(chan struct{}),
+		startedSignal:          make(chan struct{}),
+		mainWg:                 &sync.WaitGroup{},
+	}
+}
+
+func (s *Supervisor) START() {
+	s.mainWg.Add(1)
+	go s.waitForClientOperations()
+	s.IsRunning = true
+	close(s.startedSignal)
+	s.mainWg.Wait()
+	s.IsRunning = false
+
+}
+
+func (s *Supervisor) waitForClientOperations() {
+	defer s.mainWg.Done()
+	for {
+		select {
+		case o := <-s.clientOperationChannel:
+			switch o.Operation {
+			case Add:
+				s.addClient(o.Client)
+			case Remove:
+				s.removeClient(o.Client.Id)
+			default:
+				fmt.Println("no operation")
+			}
+		case <-s.stoppedSignal:
+			return
+		}
+	}
+
+}
+
+func (s *Supervisor) addClient(c *Client) error {
+	s.clientsMutex.Lock()
+	defer s.clientsMutex.Unlock()
+	s.clientsMapping[c.Id] = c
```

Comment:
question: I think later we need to verify if the id does not exist, what do you think?

Reply:
Totally true, we could try access the map and avoid the allocation if not necessary.